### PR TITLE
cudaPackages.nccl: 2.18.3-1 -> 2.18.5-1; add updateScript

### PR DESCRIPTION
**Pull request by ConnorBaker**
## Description of changes

- switched to `pname`
- added `updateScript`
- 2.18.3-1 -> 2.18.5-1
  - Used the update script to perform the update!

@NixOS/cuda-maintainers @mdaiter @orivej-nixos

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
Passed `nccl-test-suite` at https://github.com/connorbaker/nix-cuda-test.

<details>

```console
$ nix run .#nccl-test-suite --override-input nixpkgs github:nixos/nixpkgs/5a5d095526798afff81da9b8ad5c6f0df4dac313
warning: not writing modified lock file of flake 'git+file:///home/connorbaker/nix-cuda-test':
• Updated input 'nixpkgs':
    'github:NixOS/nixpkgs/cf795c35039b544a66a29a9ad3cd4d37bd1d6cb3' (2023-08-21)
  → 'github:nixos/nixpkgs/5a5d095526798afff81da9b8ad5c6f0df4dac313' (2023-08-27)
Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/all_gather_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272738 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float    none      -1    18.48  1816.16    0.00      0     0.05  619657.10    0.00      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/all_reduce_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272750 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float     sum      -1    18.46  1817.72    0.00      0     0.05  627772.35    0.00      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/alltoall_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272762 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float    none      -1    23.10  1452.73    0.00      0     5.07  6620.58    0.00    N/A
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/broadcast_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272774 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float    none       0    18.47  1816.32  1816.32      0     0.07  499321.90  499321.90      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 250569 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/gather_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272786 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float    none       0    23.28  1441.08    0.00      0     4.98  6743.39    0.00      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/hypercube_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272798 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float              -1    18.48    0.00    0.00      0     0.04    0.00    0.00      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/reduce_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272810 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float     sum       0    18.55  1808.47  1808.47      0     0.06  560642.14  560642.14      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 281225 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/reduce_scatter_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272822 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float     sum      -1    18.55  1808.56    0.00      0     0.06  563466.53    0.00      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/scatter_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272834 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float    none       0    23.01  1457.95    0.00      0     5.02  6679.23    0.00      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 0 
#

Running /nix/store/6v92dmg26nf71b46b20q4664b3d9y08i-nccl-tests-2.13.6/bin/sendrecv_perf
# nThread 1 nGpus 1 minBytes 33554432 maxBytes 33554432 step: 1048576(bytes) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#  Rank  0 Group  0 Pid 272846 on nixos-desktop device  0 [0x01] NVIDIA GeForce RTX 4090
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
    33554432       8388608     float     sum      -1    23.00  1458.70  1458.70      0     5.03  6672.19  6672.19    N/A
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 4065.44 
#
```

</details>

**Comment by ConnorBaker** [comment]
Result of `nixpkgs-review pr 251856 --extra-nixpkgs-config '{ allowUnfree = true; cudaSupport = true; cudaCapabilities = [ "8.9" ]; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorflowWithCuda.dist</li>
    <li>python311Packages.tensorflowWithCuda</li>
    <li>python311Packages.tensorflowWithCuda.dist</li>
  </ul>
</details>
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>cudaPackages.nccl</li>
    <li>cudaPackages.nccl-tests</li>
    <li>cudaPackages.nccl.dev</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.cupy.dist</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.jaxlibWithCuda.dist</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python310Packages.torchWithCuda.dev</li>
    <li>python310Packages.torchWithCuda.dist</li>
    <li>python310Packages.torchWithCuda.lib</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.cupy.dist</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.jaxlibWithCuda.dist</li>
    <li>python311Packages.torchWithCuda</li>
    <li>python311Packages.torchWithCuda.dev</li>
    <li>python311Packages.torchWithCuda.dist</li>
    <li>python311Packages.torchWithCuda.lib</li>
  </ul>
</details>

File: pkgs/development/libraries/science/math/nccl/default.nix [hunk 1]
```diff
@@ -63,6 +65,12 @@ backendStdenv.mkDerivation (finalAttrs: {
 
   env.NIX_CFLAGS_COMPILE = toString [ "-Wno-unused-function" ];
 
+  # Run the update script with: `nix-shell maintainers/scripts/update.nix --argstr package cudaPackages.nccl`
+  passthru.updateScript = gitUpdater {
+    inherit (finalAttrs) pname version;
+    rev-prefix = "v";
+  };
```
Comment (samuela) [inline pkgs/development/libraries/science/math/nccl/default.nix:72]: nice! hadn't seen this before!

**Review by samuela — APPROVED** [review]
diff LGTM

**Pull request merged**